Homeworks/02/src/solutions.js

```diff
@@ -2,88 +2,151 @@ module.exports = {
   /**
    * Returns an empty object without prototype. There is object creation type that creates object without prototype
    */
-  createPrototypelessObject() {},
+  createPrototypelessObject() {
+    return Object.create(null)
+  },
 
   /**
    * Returns an object with prototype set to given `proto`.
    * @param {Object} proto Prototype object
    */
-  createObjectWithPrototype(proto) {},
+  createObjectWithPrototype(proto) {
+    return Object.create(proto)
+  },
 
   /**
    * Returns an object with `value` property set to the given `value` and `getValue` method.
    * Be careful, if `value` changes, `getValue` should return changed `value`.
    * @param {any} value
    */
-  createObjectWithMethod(value) {},
+  createObjectWithMethod(value) {
+    return {
+      value: value,
+      getValue: function() {
+        return this.value
+      }
+    }
+  },
 
   /**
    * Returns an object with the `getValue` and `setValue` methods, having `value` hidden from the outside.
    */
-  createEncapsulatedObject() {},
+  createEncapsulatedObject() {
+    let value;
+
+    return {
+      getValue: function() {
+        return value
+      },
+      setValue: function(newValue) {
+        value = newValue
+      } 
+    }
+  },
 
   /**
    * Returns the shallow copy of the given `obj`. HINT: This **operator** will be used later.
    * @param {Object} obj
    */
-  shallowCopy(obj) {},
+  shallowCopy(obj) {
+    return Object.assign({}, obj)
+  },
 
   /**
    * Returns the deep copy of the given `obj`.
    * @param {Object} obj
    */
-  deepCopy(obj) {},
+  deepCopy(obj) {
+    return JSON.parse(JSON.stringify(obj))
```

Comment:
Ovo je oke, ali smatra se "hackom" jer nije moguće serijalizirati svaki tip podataka (klase, funkcije, date itd.) a isto tako ima slabe performanse kod većih objekata. 
Preciznije bi bilo koristiti `structuredClone` API:
https://developer.mozilla.org/en-US/docs/Glossary/Deep_copy
https://developer.mozilla.org/en-US/docs/Web/API/Window/structuredClone
```suggestion
    return structuredClone(obj);
```